src/backend/distributed/commands/table.c

```diff
@@ -666,6 +705,107 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand,
+			{
+				if (constraint->contype == CONSTR_DEFAULT)
+				{
+					if (constraint->raw_expr != NULL)
+					{
+						ParseState *pstate = make_parsestate(NULL);
+						Node *expr = transformExpr(pstate, constraint->raw_expr,
+												   EXPR_KIND_COLUMN_DEFAULT);
+
+						if (contain_nextval_expression_walker(expr, NULL) ||
+							contain_funcexpr_walker(expr, NULL))
+						{
+							deparseAT = true;
+							newCmd->subtype = command->subtype;
+
+							ColumnDef *newColDef = makeNode(ColumnDef);
+							newColDef->colname = columnDefinition->colname;
+							newColDef->typeName = columnDefinition->typeName;
+							newColDef->is_not_null = true;
+							newColDef->collClause = columnDefinition->collClause;
+							newColDef->collOid = columnDefinition->collOid;
+
+							newCmd->def = (Node *) newColDef;
+						}
+					}
+				}
+			}
+
+			/*
+			 * We check for ADD COLUMN .. serial_prototype
+			 * if that's the case, we should deparse the statement
+			 * The structure of this check is copied from transformColumnDefinition.
+			 */
+			if (columnDefinition->typeName && list_length(
+					columnDefinition->typeName->names) == 1 &&
+				!columnDefinition->typeName->pct_type)
+			{
+				char *typeName = strVal(linitial(columnDefinition->typeName->names));
+
+				if (strcmp(typeName, "smallserial") == 0 ||
+					strcmp(typeName, "serial2") == 0 ||
+					strcmp(typeName, "serial") == 0 ||
+					strcmp(typeName, "serial4") == 0 ||
+					strcmp(typeName, "bigserial") == 0 ||
+					strcmp(typeName, "serial8") == 0)
+				{
+					deparseAT = true;
+					newCmd->subtype = command->subtype;
+
+					ColumnDef *newColDef = makeNode(ColumnDef);
+					newColDef->colname = columnDefinition->colname;
+					newColDef->typeName = columnDefinition->typeName;
+					newColDef->is_not_null = true;
+					newColDef->collClause = columnDefinition->collClause;
+					newColDef->collOid = columnDefinition->collOid;
+
+					if (strcmp(typeName, "smallserial") == 0 ||
+						strcmp(typeName, "serial2") == 0)
+					{
+						newColDef->typeName->names = NIL;
+						newColDef->typeName->typeOid = INT2OID;
+					}
+					else if (strcmp(typeName, "serial") == 0 ||
+							 strcmp(typeName, "serial4") == 0)
+					{
+						newColDef->typeName->names = NIL;
+						newColDef->typeName->typeOid = INT4OID;
+					}
+					else if (strcmp(typeName, "bigserial") == 0 ||
+							 strcmp(typeName, "serial8") == 0)
+					{
+						newColDef->typeName->names = NIL;
+						newColDef->typeName->typeOid = INT8OID;
+					}
+					newCmd->def = (Node *) newColDef;
+				}
+			}
+		}
+		/*
+		 * We check for ALTER COLUMN .. SET/DROP DEFAULT
+		 * we should not propagate anything to shards
+		 */
+		else if (alterTableType == AT_ColumnDefault)
+		{
+			ParseState *pstate = make_parsestate(NULL);
+			Node *expr = transformExpr(pstate, command->def,
+									   EXPR_KIND_COLUMN_DEFAULT);
+
+			if (contain_nextval_expression_walker(expr, NULL) || contain_funcexpr_walker(
+					expr, NULL))
```

Comment:
> contain_funcexpr_walker

One thing to note is that we do include `DEFAULT function_call()` when creating shards during `create_distributed_table`. I wonder whether we should propagate it to be consistent. On the other hand, I don't know whether there is ever a case in which we use it.

It's worth noting that functions in Postgres can be volatile, stable or immutable, and they have different behaviour. E.g. INSERT..SELECT does different things for each.

Reply:
I have included this `contain_funcexpr_walker` check during `create_distributed_table` as well, but as discussed, I will not keep the function changes in this PR but rather address them later in another one. Thanks.